pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -969,50 +976,42 @@ protected void internalUnloadTopic(AsyncResponse asyncResponse, boolean authorit
-                .thenCompose(topic -> topic.close(false))
-                .thenRun(() -> {
-                    log.info("[{}] Successfully unloaded topic {}", clientAppId(), topicName);
-                    asyncResponse.resume(Response.noContent().build());
-                })
+    private void internalUnloadNonPartitionedTopicAsync(AsyncResponse asyncResponse, boolean authoritative) {
+        validateTopicOperationAsync(topicName, TopicOperation.UNLOAD)
+                .thenCompose(unused -> validateTopicOwnershipAsync(topicName, authoritative)
+                        .thenCompose(__ -> getTopicReferenceAsync(topicName))
+                        .thenCompose(topic -> topic.close(false))
+                        .thenRun(() -> {
+                            log.info("[{}] Successfully unloaded topic {}", clientAppId(), topicName);
+                            asyncResponse.resume(Response.noContent().build());
+                        }))
                 .exceptionally(ex -> {
-                    log.error("[{}] Failed to unload topic {}, {}", clientAppId(), topicName, ex.getMessage());
-                    asyncResponse.resume(ex.getCause());
+                    Throwable cause = ex.getCause();
+                    log.error("[{}] Failed to unload topic {}, {}", clientAppId(), topicName, cause);
+                    resumeAsyncResponseExceptionally(asyncResponse, cause);
                     return null;
                 });
     }
 
-    private void internalUnloadTransactionCoordinator(AsyncResponse asyncResponse, boolean authoritative) {
-        try {
-            validateTopicOperation(topicName, TopicOperation.UNLOAD);
-        } catch (Exception e) {
-            log.error("[{}] Failed to unload tc {},{}", clientAppId(), topicName.getPartitionIndex(), e.getMessage());
-            resumeAsyncResponseExceptionally(asyncResponse, e);
-            return;
-        }
-        validateTopicOwnershipAsync(topicName, authoritative)
-                .thenCompose(v -> pulsar()
-                .getTransactionMetadataStoreService()
-                .removeTransactionMetadataStore(TransactionCoordinatorID.get(topicName.getPartitionIndex())))
-                .thenRun(() -> {
-                    log.info("[{}] Successfully unloaded tc {}", clientAppId(), topicName.getPartitionIndex());
-                    asyncResponse.resume(Response.noContent().build());
-                }).exceptionally(ex -> {
-                    log.error("[{}] Failed to unload tc {}, {}", clientAppId(), topicName.getPartitionIndex(),
-                    ex.getMessage());
-            asyncResponse.resume(ex.getCause());
-            return null;
-        });
+    private void internalUnloadTransactionCoordinatorAsync(AsyncResponse asyncResponse, boolean authoritative) {
+        validateTopicOperationAsync(topicName, TopicOperation.UNLOAD)
+                .thenAccept(__ -> validateTopicOwnershipAsync(topicName, authoritative)
```

Comment:
```suggestion
                .thenCompose(__ -> validateTopicOwnershipAsync(topicName, authoritative)
```

Reply:
@codelipenghui 

I have a small question, why we use ``thenCompose`` instead of ``thenAccept`` ?

Reply:
@mattisonchao Just an example:

```java
        CompletableFuture future = new CompletableFuture();
        future.thenCompose(f -> {
            CompletableFuture e = new CompletableFuture();
            e.completeExceptionally(new RuntimeException());
            return e;
        }).exceptionally(ex -> {
            System.out.println(ex);
            return null;
        });
        future.complete(null);
```

You will see `java.util.concurrent.CompletionException: java.lang.RuntimeException`

If change to 

```java
        CompletableFuture future = new CompletableFuture();
        future.thenAccept(f -> {
            CompletableFuture e = new CompletableFuture();
            e.completeExceptionally(new RuntimeException());
        }).exceptionally(ex -> {
            System.out.println(ex);
            return null;
        });
        future.complete(null);
```

You will not get the exception.

Here you are calling an async method in `thenAccept`, so the behavior is it will move to the next stage immediately, no matter if the `validateTopicOwnershipAsync` returned a failed future or not.

Reply:
Oh, i see. it's my misunderstanding for ``thenAccept``. I need to check the source code of ``CompletableFuture``  for other misunderstandings.

@codelipenghui 
Very thanks for your good example code and explain patiently.